### ISSUE TITLE
NavigationStarting event is not deferrable

### ISSUE THREAD

**Issue by nnemkin** [MicrosoftEdge/WebViewFeedback#126]
`NavigationStarting` and `FrameNavigationStarting` events should support deferral, since they have an output argument (`put_Cancel`).

**Comment by david-risney** [comment]
Thanks for your feedback! I agree although it is actually a lot of work to make that happen for NavigationStarting. Were you just noting that events with side effects or out parameters should be deferrable, or do you have a scenario where you need NavigationStarting to be deferrable? Can you elaborate on how you will use a deferrable NavigationStarting? This will help us prioritize this work. Thanks!

**Comment by nnemkin** [comment]
I'm adding WebView2 backend to SWT Browser control (tracking bug here: https://bugs.eclipse.org/bugs/show_bug.cgi?id=538991). SWT is a cross-platform Java GUI toolkit, the basis of Eclipse Platform and IDE. Very roughly speaking, it's like WinForms in Java.

SWT Browser API is almost fully synchronous (except Navigate), while most WebView2 APIs are async. To preserve API compatibility, I have to fake synchronicity by spinning nested message loops, waiting for WebView2 completion handlers to fire. This breaks down when waiting happens inside an event handler, because WebView2 events are serialized. While an event is processed, no other events or completion handlers are delivered.

For example, it's impossible to create another WebView2 from inside of NavigationStarting event: create completion won't fire until NavigationStarting returns, but that's the kind of functionality SWT must provide. Another major source of reentrancy issues is ExecuteScript. SWT clients want to run scripts with return values from event handlers. Finally, there are synchronous RemoteObject callbacks (I'm not sure if they block completions or not).

An obvious work-around to unblock completion delivery is to defer event processing, which leads us to this feature request. It doesn't solve all reentrancy problems, but it's the best option available so far.

In short, what I actually need is not deferrable events, but synchronous versions of `CreateCoreWebView2EnvironmentWithDetails`, `CreateCoreWebView2Host` and `ExecuteScript`, freely callable from any context, be it an event handler or remote object callback.

**Comment by nnemkin** [comment]
Forgot to add: eventually, SWT will remove APIs that require synchronous ExecuteScript and synchronous RemoteObject callbacks, because other browser components (WebKitGTK, WKWebView) don't support them either. But synchronous Create functions would be very helpful in any case.
